clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+     * transition to the {@link MemberState#LEAVING} state so that a heartbeat
+     * request is sent out with it.
+     */
+    private void transitionToSendingLeaveGroup() {
+        memberEpoch = ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH;
+        currentAssignment = new HashSet<>();
+        transitionTo(MemberState.LEAVING);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldHeartbeatNow() {
+        MemberState state = state();
+        return state == MemberState.ACKNOWLEDGING || state == MemberState.LEAVING;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onHeartbeatRequestSent() {
+        MemberState state = state();
+        if (state == MemberState.ACKNOWLEDGING) {
+            if (allPendingAssignmentsReconciled()) {
+                transitionTo(MemberState.STABLE);
+            } else {
+                transitionTo(MemberState.RECONCILING);
+            }
+        } else if (state == MemberState.LEAVING) {
+            transitionTo(MemberState.UNSUBSCRIBED);
         }
-        return state.equals(MemberState.STABLE);
     }
 
     /**
-     * Take new target assignment received from the server and set it as targetAssignment to be
-     * processed. Following the consumer group protocol, the server won't send a new target
-     * member while a previous one hasn't been acknowledged by the member, so this will fail
-     * if a target assignment already exists.
+     * @return True if there are no assignments waiting to be resolved from metadata or reconciled.
+     */
+    private boolean allPendingAssignmentsReconciled() {
+        return assignmentUnresolved.isEmpty() && assignmentReadyToReconcile.isEmpty();
+    }
+
+    @Override
+    public boolean shouldSkipHeartbeat() {
+        MemberState state = state();
+        return state == MemberState.UNSUBSCRIBED || state == MemberState.FATAL;
+    }
+
+    /**
+     * Reconcile the assignment that has been received from the server and for which topic names
+     * are resolved, kept in the {@link #assignmentReadyToReconcile}. This will commit if needed,
+     * trigger the callbacks and update the subscription state. Note that only one reconciliation
+     * can be in progress at a time. If there is already another one in progress when this is
+     * triggered, it will be no-op, and the assignment will be reconciled on the next
+     * reconciliation loop.
+     */
+    boolean reconcile() {
+        // Make copy of the assignment to reconcile as it could change as new assignments or metadata updates are received
+        SortedSet<TopicPartition> assignedPartitions = new TreeSet<>(COMPARATOR);
+        assignedPartitions.addAll(assignmentReadyToReconcile);
+
+        SortedSet<TopicPartition> ownedPartitions = new TreeSet<>(COMPARATOR);
+        ownedPartitions.addAll(subscriptions.assignedPartitions());
+
+        boolean sameAssignmentReceived = assignedPartitions.equals(ownedPartitions);
```

Comment:
Is this really true? We could have the same topic name in both but with different topic ids for instance. In my opinion, we should move towards using TopicIdPartition for both the assigned partitions and the partitions ready to reconcile. We can of course tackle separately from this PR.

Reply:
Agree, filed [KAFKA-15839](https://issues.apache.org/jira/browse/KAFKA-15839) to extend topic ID usage in the whole assignment reconciliation flow to make sure we handle topic re-creation properly.